Cat_vs_Dog_Classification_Using_Deep_Learning

In [ ]:
import tensorflow as tf

In [2]:
tf.test.is_gpu_available()

Instructions for updating:
Use `tf.config.list_physical_devices('GPU')` instead.


True

In [3]:
print("Num GPUs Available: ", len(tf.config.experimental.list_physical_devices('GPU')))
#Num GPUs Available: 1
tf.test.is_built_with_cuda()
#True
print(tf.version. VERSION)
#2.1.0


Num GPUs Available:  1
2.10.1


In [4]:
!python --version


Python 3.10.14


In [5]:
from tensorflow import keras
from keras import Sequential

In [6]:
from keras.layers import Dense,Conv2D,MaxPooling2D,Flatten

In [7]:
model = Sequential()

In [8]:
#this is my first convololution layer 
model.add(Conv2D(32,kernel_size=(3,3),padding="valid",activation="relu",input_shape=(256,256,3)))
model.add(MaxPooling2D(pool_size=(2,2),strides=2,padding='valid'))
#second convololution
model.add(Conv2D(64,kernel_size=(3,3),padding="valid",activation="relu"))
model.add(MaxPooling2D(pool_size=(2,2),strides=2,padding='valid'))
#third convololution
model.add(Conv2D(128,kernel_size=(3,3),padding="valid",activation="relu"))
model.add(MaxPooling2D(pool_size=(2,2),strides=2,padding='valid')) 
#adding first hidden layer with 128 neurons and relu activation
model.add(Flatten())
model.add(Dense(128,activation='relu'))
#adding second hidden layer with 64 neurons and relu activation
model.add(Dense(64,activation='relu'))
#here we have a 1 output since it is a binary classification
model.add(Dense(1,activation='sigmoid'))

In [9]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 254, 254, 32)      896       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 127, 127, 32)     0         
 )                                                               
                                                                 
 conv2d_1 (Conv2D)           (None, 125, 125, 64)      18496     
                                                                 
 max_pooling2d_1 (MaxPooling  (None, 62, 62, 64)       0         
 2D)                                                             
                                                                 
 conv2d_2 (Conv2D)           (None, 60, 60, 128)       73856     
                                                                 
 max_pooling2d_2 (MaxPooling  (None, 30, 30, 128)      0

In [10]:
model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

Now comes the biggest challenge , giving data to model

In [11]:
#internally it is using generator concept which is there in python , we know we use 
#generator concept when we need memory efficient way to load something not directly
train_ds = keras.utils.image_dataset_from_directory(
    directory=r"C:/Users/anilk/Desktop/test",
    batch_size=32,
    image_size=(256,256), #resytriction in image
    labels="inferred",
    label_mode="int"
    
)
validation_ds =keras.utils.image_dataset_from_directory(
    directory=r'C:/Users/anilk/Desktop/train',
    batch_size=32,
    image_size=(256,256),
    labels="inferred",
    label_mode="int"
)

Found 25000 files belonging to 2 classes.
Found 4519 files belonging to 2 classes.


In our dataset there are many photos , and their sizes are also very different 

if i want to pr

in ML we do standardization and normalisation 

but in Dl we do min max Scaling (pixcel scaling) , unit scaling

pixcel size is [0-255] , image size is 256*256 , if i want to compress pixcell size we have to  divide it by 255, then value is cmpressed between [0-1]

In [12]:
def process(image,label):
    tf.cast(image/255,tf.float32)
    return image,label
train_ds = train_ds.map(process)
validation_ds = validation_ds.map(process)

In [13]:
import tensorflow as tf

physical_devices = tf.config.experimental.list_physical_devices('GPU')
print("Number of available GPUs:", len(physical_devices))

for device in physical_devices:
    print(device.name, tf.config.experimental.get_memory_info(device))


Number of available GPUs: 1


TypeError: TFE_GetMemoryInfo(): incompatible function arguments. The following argument types are supported:
    1. (arg0: handle, arg1: str) -> Dict[str, int]

Invoked with: <capsule object NULL at 0x0000014DFB774D20>, PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')

In [14]:
model.fit(train_ds,epochs=10,validation_data=validation_ds)

Epoch 1/10
782/782 [==============================] - 119s 137ms/step - loss: 3.1658 - accuracy: 0.5889 - val_loss: 0.6615 - val_accuracy: 0.5990
Epoch 2/10
782/782 [==============================] - 106s 135ms/step - loss: 0.6708 - accuracy: 0.5927 - val_loss: 0.6454 - val_accuracy: 0.6156
Epoch 3/10
782/782 [==============================] - 107s 137ms/step - loss: 0.6205 - accuracy: 0.6549 - val_loss: 0.7176 - val_accuracy: 0.5966
Epoch 4/10
782/782 [==============================] - 102s 130ms/step - loss: 0.6338 - accuracy: 0.6243 - val_loss: 0.5922 - val_accuracy: 0.6734
Epoch 5/10
782/782 [==============================] - 99s 127ms/step - loss: 0.5328 - accuracy: 0.7062 - val_loss: 0.6062 - val_accuracy: 0.7123
Epoch 6/10
782/782 [==============================] - 100s 128ms/step - loss: 0.4576 - accuracy: 0.7606 - val_loss: 0.7663 - val_accuracy: 0.6917
Epoch 7/10
782/782 [==============================] - 105s 134ms/step - loss: 0.3724 - accuracy: 0.8134 - val_loss: 0.8545 - 

In [16]:
!nvidia-smi

/bin/bash: line 1: nvidia-smi: command not found


In [ ]:
!pip install --upgrade tensorflow

Defaulting to user installation because normal site-packages is not writeable


In [44]:
import numpy as np

In [ ]:
model.predict()

In [49]:
loaded_array = np.loadtxt("C:/Users/anilk/Downloads/image_array.txt").reshape(1,256, 256, 3)

model.save("C:\Users\anilk\Desktop\model")

In [28]:
model.save("C:/Users/anilk/Desktop/model")

INFO:tensorflow:Assets written to: C:/Users/anilk/Desktop/model\assets


INFO:tensorflow:Assets written to: C:/Users/anilk/Desktop/model\assets


In [30]:
model.save_weights("C:/Users/anilk/Desktop/model2")

In [36]:

model.save_weights('C:/Users/anilk/Desktop/my_weights')


In [52]:
a= model.predict(loaded_array)

1/1 [==============================] - 0s 51ms/step


In [53]:
print(a)

[[0.5974293]]


In [54]:
loaded_array = np.loadtxt("C:/Users/anilk/Downloads/image_array2.txt").reshape(1,256, 256, 3)

In [55]:
a1 = model.predict(loaded_array)

1/1 [==============================] - 0s 91ms/step


In [56]:
print(a1)

[[0.99989235]]


In [57]:
loaded_array = np.loadtxt("C:/Users/anilk/Downloads/image_array3.txt").reshape(1,256, 256, 3)
a2 = model.predict(loaded_array)
print(a2)

1/1 [==============================] - 0s 41ms/step
[[0.6362219]]


In [59]:
loaded_array = np.loadtxt("C:/Users/anilk/Downloads/image_array4.txt").reshape(1,256, 256, 3)
a3 = model.predict(loaded_array)
print(a3)

1/1 [==============================] - 0s 64ms/step
[[0.6443196]]


In [60]:
loaded_array = np.loadtxt("C:/Users/anilk/Downloads/image_array5.txt").reshape(1,256, 256, 3)
a5 = model.predict(loaded_array)
print(a5)

1/1 [==============================] - 0s 50ms/step
[[0.97005916]]


In this project i have faced many issue such that , 

Issue 1 : i have dream of training neural networks on my local system GPU , so i have faces multiple issues in that , i have windows 11 machine , and is 2024 year and tensorflow with GPU support is not supported in tensorflowversion greater than 2.10 , and for python version greater than 3.10 , so i have to first handle taht issue , and also i have to switch from jupyter notebook to VS code to run my model on GPU , due to vS code i am facing issue with matplotlib , but i handled that issue as well 

Issue 2: i am unable to read any image array in my vscode, i tried different libraries for that like Cv2, tensorflow.preprocessing as well, but i explictly generated image array of that specific image that i want to test with using given code , 

In [ ]:
"""# Path to your image
image_path = "/content/dog.67.jpg"

# Read the image and resize it to 256x256 during loading
img = load_img(image_path, target_size=(256, 256))

# Convert the image to a NumPy array
img_array = img_to_array(img)

# Expand dimensions for a batch of one image
img_array = np.expand_dims(img_array, axis=0)

# Remove the batch dimension for saving
img_array = np.squeeze(img_array, axis=0)

# Save the array to a text file
np.savetxt("image_array5.txt", img_array.reshape(-1, img_array.shape[-1]), fmt='%d')

# Alternatively, save the array to a binary file for more compact storage
np.save("image_array.npy", img_array)"""

# or to check right image array is generated or not 

""" import numpy as np

# Load the array from the text file
loaded_array = np.loadtxt("/content/image_array5.txt").reshape(256, 256, 3)

# Display the image to verify it is loaded correctly
import matplotlib.pyplot as plt

plt.imshow(loaded_array.astype('uint8'))
plt.axis('off')  # Hide axis
plt.show()
"""

Issue 3 : from where i download dataset , here in train folder  , dog and cat images are in same , so i have to write script to make different folders of dog and cat , and in test folder all iamges are not caategoried * their names is also not specified , so i have to also write script to make validation dataset

kaggle link where i downlaod dataset: https://www.kaggle.com/c/dogs-vs-cats/data

In [ ]:
# script to move photos to make 2 folders in train folder named cat and dog 
"""import os
import shutil

# Source directory containing dog and cat images
source_dir = "C:/Users/anilk/Downloads/dogs-vs-cats/test1/test1"  # Replace with your actual directory path

# Destination directories for sorted images
dog_dir = "C:/Users/anilk/Desktop/train\dog"
cat_dir = "C:/Users/anilk/Desktop/train\cat"

# Create the destination directories if they don't exist
if not os.path.exists(dog_dir):
    os.makedirs(dog_dir)

if not os.path.exists(cat_dir):
    os.makedirs(cat_dir)

# Move or copy images based on their filename pattern
for filename in os.listdir(source_dir):
    if filename.startswith("cat"):  # Move cat images to cat_dir
        shutil.move(os.path.join(source_dir, filename), os.path.join(cat_dir, filename))
    elif filename.startswith("dog"):  # Copy dog images to dog_dir
        shutil.copy(os.path.join(source_dir, filename), os.path.join(dog_dir, filename))"""


# script to make validation dataset from training dataset 

"""import os
import random
import shutil

# Source directories
train_dog_dir =  "c:/Users/anilk/Desktop/test/dog" # Replace with actual paths "C:\Users\anilk\Desktop\train\cat"
train_cat_dir = "c:/Users/anilk/Desktop/test/cat"  # Replace with actual paths

# Destination directories
test_dog_dir =  "c:/Users/anilk/Desktop/train/dog" # Replace with actual paths "C:\Users\anilk\Desktop\train\cat"
test_cat_dir =  "c:/Users/anilk/Desktop/train/cat" # Replace with actual paths

# Number of images to copy per category (dog and cat)
num_images_per_category = 2500
print(2)
# Copy random images from train_dog_dir to test_dog_dir
for _ in range(num_images_per_category):
    # Get a random image filename from train_dog_dir
    filename = random.choice(os.listdir(train_dog_dir))

    # Create the destination path for the image in test_dog_dir
    dest_path = os.path.join(test_dog_dir, filename)

    # Copy the image from source to destination
    shutil.copy(os.path.join(train_dog_dir, filename), dest_path)
    print(9)
# Copy random images from train_cat_dir to test_cat_dir
for _ in range(num_images_per_category):
    # Get a random image filename from train_cat_dir
    filename = random.choice(os.listdir(train_cat_dir))

    # Create the destination path for the image in test_cat_dir
    dest_path = os.path.join(test_cat_dir, filename)

    # Copy the image from source to destination
    shutil.copy(os.path.join(train_cat_dir, filename), dest_path)
 """